akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala

```diff
@@ -0,0 +1,139 @@
+    }
+    val nextGen = timerGen.next()
+
+    val timerMsg = TimerMsg(key, nextGen, this)
+    val task =
+      if (repeat)
+        ctx.system.scheduler.schedule(timeout, timeout) {
+          ctx.self.upcast ! timerMsg
+        }(ExecutionContexts.sameThreadExecutionContext)
+      else
+        ctx.system.scheduler.scheduleOnce(timeout) {
+          ctx.self.upcast ! timerMsg
+        }(ExecutionContexts.sameThreadExecutionContext)
+
+    val nextTimer = Timer(key, msg, repeat, nextGen, task)
+    log.debug("Start timer [{}] with generation [{}]", key, nextGen)
+    timers = timers.updated(key, nextTimer)
+  }
+
+  override def isTimerActive(key: Any): Boolean =
+    timers.contains(key)
+
+  override def cancel(key: Any): Unit = {
+    timers.get(key) match {
+      case None    ⇒
+      case Some(t) ⇒ cancelTimer(t)
+    }
+  }
+
+  private def cancelTimer(timer: Timer[T]): Unit = {
+    log.debug("Cancel timer [{}] with generation [{}]", timer.key, timer.generation)
+    timer.task.cancel()
+    timers -= timer.key
+  }
+
+  override def cancelAll(): Unit = {
+    log.debug("Cancel all timers")
+    timers.valuesIterator.foreach { timer ⇒
+      timer.task.cancel()
+    }
+    timers = Map.empty
+  }
+
+  private def interceptTimerMsg(ctx: ActorContext[TimerMsg], timerMsg: TimerMsg): T = {
+    timers.get(timerMsg.key) match {
+      case None ⇒
+        // it was from canceled timer that was already enqueued in mailbox
+        log.debug("Received timer [{}] that has been removed, discarding", timerMsg.key)
+        null.asInstanceOf[T] // message should be ignored
+      case Some(t) ⇒
+        if (timerMsg.owner ne this) {
+          // after restart, it was from an old instance that was enqueued in mailbox before canceled
+          log.debug("Received timer [{}] from old restarted instance, discarding", timerMsg.key)
+          null.asInstanceOf[T] // message should be ignored
+        } else if (timerMsg.generation == t.generation) {
+          // valid timer
+          log.debug("Received timer [{}]", timerMsg.key)
+          if (!t.repeat)
+            timers -= t.key
+          t.msg
+        } else {
+          // it was from an old timer that was enqueued in mailbox before canceled
+          log.debug(
+            "Received timer [{}] from from old generation [{}], expected generation [{}], discarding",
+            timerMsg.key, timerMsg.generation, t.generation)
+          null.asInstanceOf[T] // message should be ignored
+        }
+    }
+  }
+
+  def intercept(behavior: Behavior[T]): Behavior[T] = {
+    // The scheduled TimerMsg is intercepted to guard against old messages enqueued
+    // in mailbox before timer was canceled.
+    // Intercept some signals to cancel timers when when restarting and stopping.
+    BehaviorImpl.Intercept[T, TimerMsg](
+      beforeOnMessage = interceptTimerMsg,
+      beforeOnSignal = (ctx, sig) ⇒ {
+      sig match {
+        case PreRestart | PostStop ⇒ cancelAll()
```

Comment:
Hah! Very nice, biggest improvements of the entire PR to timers in actors in general I think :)